### PR TITLE
Adding Hashicorp vault support

### PR DESCRIPTION
**Pull request by vnzongzna**
Followup #623

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -290,6 +290,66 @@ And decrypt it using::
 	 $ sops --decrypt test.enc.yaml
 
 
+Encrypting using Hashicorp Vault
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To understand how to deploy Vault securely this is not the place, we assume you have a instance (or more) of Vault running and you have privileged access to it.
```
Comment (autrilla) [inline README.rst:296]:
```suggestion
We assume you have an instance (or more) of Vault running and you have privileged access to it. For instructions on how to deploy a secure instance of Vault, refer to Hashicorp's official documentation.
```

File: README.rst [hunk 2]
```diff
@@ -290,6 +290,66 @@ And decrypt it using::
...
+To easily deploy Vault locally: (DO NOT DO THIS FOR PRODUCTION!!!) 
+
+.. code:: bash
+
+	$ docker run -d -p8200:8200 vault:1.2.0 server -dev -dev-root-token-id=toor
+
+
+.. code:: bash
+
+	$ # Subsitute this with the address of the vault running
```
Comment (autrilla) [inline README.rst:307]:
```suggestion
	$ # Substitute this with the address Vault is running on
```

File: README.rst [hunk 3]
```diff
@@ -290,6 +290,66 @@ And decrypt it using::
...
+	$ export VAULT_ADDR=http://127.0.0.1:8200 
+
+	$ # this may not be necessary in case you previously used `vault login` for production use
+	$ export VAULT_TOKEN=toor 
+	
+	$ # to check if Vault started correctly and configured
```
Comment (autrilla) [inline README.rst:313]:
```suggestion
	$ # to check if Vault started and is configured correctly
```

File: README.rst [hunk 4]
```diff
@@ -290,6 +290,66 @@ And decrypt it using::
...
+	Total Shares    1
+	Threshold       1
+	Version         1.2.0
+	Cluster Name    vault-cluster-618cc902
+	Cluster ID      e532e461-e8f0-1352-8a41-fc7c11096908
+	HA Enabled      false
+
+	$ # It is required to enable a transit engine if not already done (It is suggested to create a transit engine specifically for sops, in which it is possible to have multiple keys with various permission levels)
+	$ vault secrets enable -path=sops transit
+	Success! Enabled thˇe transit secrets engine at: sops/
```
Comment (autrilla) [inline README.rst:329]:
```suggestion
	Success! Enabled the transit secrets engine at: sops/
```

File: README.rst [hunk 5]
```diff
@@ -290,6 +290,66 @@ And decrypt it using::
...
+
+	$ vault write sops/keys/secondkey type=rsa-2048
+	Success! Data written to: sops/keys/secondkey
+
+	$ vault write sops/keys/thirdkey type=chacha20-poly1305
+	Success! Data written to: sops/keys/thirdkey
+
+	$ sops --hc-vault-transit $VAULT_ADDR/v1/sops/keys/firstkey vault_example.yml
+
+	$ cat <<EOF >.sops.yaml
```
Comment (autrilla) [inline README.rst:343]:
```suggestion
	$ cat <<EOF > .sops.yaml
```

File: README.rst [hunk 6]
```diff
@@ -290,6 +290,66 @@ And decrypt it using::
...
+	creation_rules:
+		- path_regex: \.dev\.yaml$
+		  hc_vault_transit_uri: "$VAULT_ADDR/v1/sops/keys/secondkey"
+		- \.prod\.yaml$
```
Comment (autrilla) [inline README.rst:347]:
```suggestion
		- path_regex: \.prod\.yaml$
```

File: cmd/sops/main.go [hunk 7]
```diff
@@ -78,6 +79,14 @@ func main() {
    (you need to setup google application default credentials. See
     https://developers.google.com/identity/protocols/application-default-credentials)
 
+
+   To encrypt or decrypt a document with HashiCorp Vault's Transit Secret Engine, specify the
+   Vault key URI name in the --hc-vault-transit flag or in the SOPS_VAULT_URIS (eg. https://vault.example.org:8200/v1/transit/keys/dev
```
Comment (autrilla) [inline cmd/sops/main.go:84]:
```suggestion
   Vault key URI name in the --hc-vault-transit flag or in the SOPS_VAULT_URIS environment variable (eg. https://vault.example.org:8200/v1/transit/keys/dev
```

File: cmd/sops/main.go [hunk 8]
```diff
@@ -351,6 +360,10 @@ func main() {
 							Name:  "azure-kv",
 							Usage: "the Azure Key Vault key URL the new group should contain. Can be specified more than once",
 						},
+						cli.StringSliceFlag{
+							Name:  "hc-vault",
```
Comment (autrilla) [inline cmd/sops/main.go:364]: This and all further references should be `hc-vault-transit`, no?

File: cmd/sops/main.go [hunk 9]
```diff
@@ -983,6 +1043,8 @@ func keyGroups(c *cli.Context, file string) ([]sops.KeyGroup, error) {
 	group = append(group, cloudKmsKeys...)
 	group = append(group, azkvKeys...)
 	group = append(group, pgpKeys...)
+	group = append(group, hcVaultMkKeys...)
+	log.Infof("Master keys available:  %+v", group)
```
Comment (autrilla) [inline cmd/sops/main.go:1047]: Please remove this or make it a debug log.

File: config/config.go [hunk 10]
```diff
@@ -110,6 +118,8 @@ type creationRule struct {
 	PGP               string
 	GCPKMS            string     `yaml:"gcp_kms"`
 	AzureKeyVault     string     `yaml:"azure_keyvault"`
+	VaultURI          string     `yaml:"hc_vault_transit_uri"`
+	Vault             []vaultKey `yaml:"hc_vault_transit"`
```
Comment (autrilla) [inline config/config.go:122]: As discussed in the parent PR, we will not be support two ways of doing this. We will only support URIs or structured data, but not both, and it must be the same through CLI flags and the config.

File: hcvault/keysource.go [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+	u.Path = ""
+	return NewMasterKey(u.String(), enginePath, keyName), nil
+
+}
+
+func getBackendAndKeyFromPath(fullPath string) (enginePath, keyName string, err error) {
+	// Running vault behind a reverse proxy with longer urls seems not to be supported
+	// by the vault client api so we have a separate Error for that here.
+	if re := regexp.MustCompile(`/[^/]+/v[\d]+/[^/]+/[^/]+/[^/]+`); re.Match([]byte(fullPath)) {
+		return "", "", fmt.Errorf("Running Vault with an prefixed url is not supported! (Format has to be like https://vault.example.com:8200/v1/transit/keys/keyName)")
```
Comment (autrilla) [inline hcvault/keysource.go:84]:
```suggestion
		return "", "", fmt.Errorf("running Vault with a prefixed url is not supported! (Format has to be like https://vault.example.com:8200/v1/transit/keys/keyName)")
```

File: hcvault/keysource.go [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+	} else if re := regexp.MustCompile(`/v[\d]+/[^/]+/[^/]+/[^/]+`); re.Match([]byte(fullPath)) == false {
+		return "", "", fmt.Errorf("Vault path does not seem to be formatted correctly: (eg. https://vault.example.com:8200/v1/transit/keys/keyName)")
```
Comment (autrilla) [inline hcvault/keysource.go:86]:
```suggestion
		return "", "", fmt.Errorf("vault path does not seem to be formatted correctly: (eg. https://vault.example.com:8200/v1/transit/keys/keyName)")
```

Go errors start with lowercase letters

File: stores/dotenv/store.go [hunk 13]
```diff
@@ -81,7 +81,7 @@ func (store *Store) LoadPlainFile(in []byte) (sops.TreeBranches, error) {
 			}
 			branch = append(branch, sops.TreeItem{
 				Key:   string(line[:pos]),
-				Value: strings.Replace(string(line[pos+1:]), "\\n", "\n", -1),
+				Value: string(line[pos+1:]),
```
Comment (autrilla) [inline stores/dotenv/store.go:84]: The changes to this file don't seem to belong in this PR

File: stores/dotenv/store_test.go [hunk 14]
```diff
@@ -13,7 +13,6 @@ VAR1=val1
 VAR2=val2
 #comment
 VAR3_unencrypted=val3
-VAR4=val4\nval4
```
Comment (autrilla) [inline stores/dotenv/store_test.go:16]: Same as above

File: cmd/sops/main.go (hunk 8 above)
Comment (vnzongzna) [reply cmd/sops/main.go:364]: Yes, that is correct

File: config/config.go (hunk 10 above)
Comment (vnzongzna) [reply config/config.go:122]: Since all the other KMS are using URI, I believe going with URI would be the best choice for consistency.

File: stores/dotenv/store.go (hunk 13 above)
Comment (vnzongzna) [reply stores/dotenv/store.go:84]: It might have been a mistake, I'll undo all changes in `stores/dotenv/store.go`

File: cmd/sops/main.go (hunk 8 above)
Comment (autrilla) [reply cmd/sops/main.go:364]: Can you change it please? It still shows `hc-vault`.

File: config/config.go (hunk 10 above)
Comment (autrilla) [reply config/config.go:122]: That sounds like the best option to me right now, yes.

**Comment by FWest98** [comment]
Is there any information on when we can expect this in a release? For now, I am manually compiling the binaries in my docker file, but that is not ideal.

**Comment by l0nax** [comment]
Yea, we are also interested that a new release is created so we can use this feature via the official binaries.

File: README.rst [hunk 15]
```diff
@@ -287,6 +287,66 @@ And decrypt it using::
...
+	$ vault write sops/keys/thirdkey type=chacha20-poly1305
+	Success! Data written to: sops/keys/thirdkey
+
+	$ sops --hc-vault-transit $VAULT_ADDR/v1/sops/keys/firstkey vault_example.yml
+
+	$ cat <<EOF > .sops.yaml
+	creation_rules:
+		- path_regex: \.dev\.yaml$
+		  hc_vault_transit_uri: "$VAULT_ADDR/v1/sops/keys/secondkey"
+		- path_regex: \.prod\.yaml$
```
Comment (seh) [inline README.rst:344]: Was this regular expression supposed to match the file paths mentioned on line 348 below?
